library/platform_util.c

```diff
@@ -143,3 +143,109 @@ struct tm *mbedtls_platform_gmtime_r( const mbedtls_time_t *tt,
 void (*mbedtls_test_hook_test_fail)( const char *, int, const char *);
 #endif /* MBEDTLS_TEST_HOOKS */
 
+#if !defined(MBEDTLS_PLATFORM_BUF_CACHE_ALT)
+typedef struct mbedtls_platform_buf_cache mbedtls_platform_buf_cache;
+struct mbedtls_platform_buf_cache
+{
+    mbedtls_platform_buf_cache *next;
+    size_t                     start;
+    size_t                     size;
+};
+
+#if defined(_MSC_VER)
+#pragma warning( push )
+/* MSVC report size_t to int and __int64 to int warning.
+ * But that is expected and safe operation.
+ */
+#pragma warning( disable : 4267 )
+#pragma warning( disable : 4244 )
+#endif
+
+int mbedtls_platform_buf_cache_write(
+        void **context, const void *buf, size_t len )
+{
+    unsigned char *p;
+    mbedtls_platform_buf_cache *tail;
+    mbedtls_platform_buf_cache *head;
+
+    if( buf == NULL || len == 0 )
+        return( MBEDTLS_ERR_SSL_BAD_INPUT_DATA );
+
+    p = mbedtls_calloc( 1, sizeof(mbedtls_platform_buf_cache) + len );
+    if( p == NULL )
+        return( MBEDTLS_ERR_SSL_ALLOC_FAILED );
+    tail = (mbedtls_platform_buf_cache *)p;
+    tail->size = len;
+
+    p += sizeof(mbedtls_platform_buf_cache);
+
+    memcpy( p, buf, len );
+
+    if( *context == NULL )
+    {
+        *context = tail;
+        return( len );
+    }
+
+    head = (mbedtls_platform_buf_cache *)*context;
+    while( head->next != NULL );
+    head->next = tail;
+
+    return( len );
+}
+
+int mbedtls_platform_buf_cache_read( void **context, void *buf, size_t len )
+{
+    mbedtls_platform_buf_cache *head;
+    mbedtls_platform_buf_cache *tmp;
+    unsigned char *p = (unsigned char *)buf;
+    unsigned char *end = p + len;
+
+    if( *context == NULL )
+    {
+        return( MBEDTLS_ERR_SSL_BAD_INPUT_DATA );
+    }
+
+    head = (mbedtls_platform_buf_cache *)*context;
+    if( buf == NULL || len == 0 )
+    {
+        while( head != NULL )
+        {
+            tmp = head;
+            head = head->next;
+            mbedtls_free( tmp );
+        }
+        *context = NULL;
+        return( 0 );
+    }
+
+    while( p < end && head != NULL )
+    {
+        unsigned char *p_start = (unsigned char *)&head[1];
+        p_start += head->start;
+        if( (size_t)(end - p) >= head->size - head->start )
```

Comment:
```suggestion
        if( (size_t)( end - p ) >= head->size - head->start )
```

Reply:
Before design review finished, I won't change it.